File: frontend/app/clock.ts
import { CONFIG } from "./config"

const { timing } = CONFIG

// prefersReducedMotion is read fresh on every call rather than cached: it can change mid-session
// (the user can flip the OS setting without reloading), and matchMedia is cheap enough that there's
// no real cost to asking again each time blink() decides.
function prefersReducedMotion(): boolean {
  return window.matchMedia("(prefers-reduced-motion: reduce)").matches
}

// GameClock tracks elapsed play time, pause state, and destination blinking cadence.
export class GameClock {
  levelDurationMs: number
  startedAt: number
  // pausedAt is a timestamp, not a flag: 0 means the clock has never been stopped or has since
  // resumed, and any other value is the instant the current pause began. isPaused is the only
  // place that sentinel is read, so no caller has to remember which way round it goes.
  pausedAt: number
  pausedDuration: number

  constructor(levelDurationMs: number) {
    this.levelDurationMs = levelDurationMs
    this.startedAt = performance.now()
    this.pausedAt = 0
    this.pausedDuration = 0
  }

  // isPaused reports whether active time is currently frozen.
  get isPaused(): boolean {
    return this.pausedAt !== 0
  }

  // elapsed reports active time only, excluding any paused duration.
  elapsed(now = performance.now()): number {
    // Freeze elapsed time while paused so reloads and manual pauses keep the remaining time stable.
    const effectiveNow = this.isPaused ? this.pausedAt : now
    return effectiveNow - this.startedAt - this.pausedDuration
  }

  // remaining returns the still-playable time clamped at zero.
  remaining(now = performance.now()): number {
    return Math.max(0, this.levelDurationMs - this.elapsed(now))
  }

  // blink toggles the destination marker in configured visibility phases. A steady, always-visible
  // marker instead of the toggle itself when the user prefers reduced motion — still findable,
  // just without the blinking that preference exists specifically to avoid.
  blink(now = performance.now()): boolean {
    if (prefersReducedMotion()) {
      return true
    }

    // Alternate the destination visibility while the round is active.
    return Math.floor(this.elapsed(now) / timing.blinkIntervalMs) % 2 === 0
  }

  // pause captures the instant at which active time should stop advancing.
  pause(now = performance.now()): void {
    // Already stopped: keep the original instant so a repeated pause cannot extend the frozen span.
    if (this.isPaused) {
      return
    }

    this.pausedAt = now
  }

  // resume folds the paused span back into the accumulated pause duration.
  resume(now = performance.now()): void {
    // Already running: there is no frozen span to fold back in.
    if (!this.isPaused) {
      return
    }

    this.pausedDuration += now - this.pausedAt
    this.pausedAt = 0
  }
}

File: maze/clock.go
package maze

import "time"

// GameClock tracks elapsed and remaining time for a level while excluding
// any span where gameplay is paused.
type GameClock struct {
	startedAt      time.Time
	pausedAt       time.Time
	pausedDuration time.Duration
	levelDuration  time.Duration
}

// NewGameClock creates a new per-level clock using the provided duration budget.
func NewGameClock(levelDuration time.Duration) GameClock {
	return GameClock{
		startedAt:     time.Now(),
		levelDuration: levelDuration,
	}
}

// Elapsed subtracts paused time so scoring and timeout logic continue from the same point after resume.
func (clock *GameClock) Elapsed() time.Duration {
	return clock.elapsedAt(time.Now())
}

// Remaining clamps at zero because timers should never be reset with a negative duration.
func (clock *GameClock) Remaining() time.Duration {
	remaining := clock.levelDuration - clock.Elapsed()
	if remaining < 0 {
		return 0
	}

	return remaining
}

// Pause records when the clock stopped advancing.
func (clock *GameClock) Pause() {
	clock.pausedAt = time.Now()
}

// Resume accumulates the paused duration and resumes elapsed-time accounting.
func (clock *GameClock) Resume() {
	if clock.pausedAt.IsZero() {
		return
	}

	// Resume accumulates the paused span instead of shifting startedAt so elapsed math stays simple.
	clock.pausedDuration += time.Since(clock.pausedAt)
	clock.pausedAt = time.Time{}
}

// BlinkOn reports whether the live goal marker should be visible on the current
// refresh tick. The blink cadence follows elapsed gameplay time, so paused time
// does not advance the animation.
func (clock *GameClock) BlinkOn() bool {
	return clock.blinkOnAt(time.Now())
}

func (clock *GameClock) elapsedAt(now time.Time) time.Duration {
	if !clock.pausedAt.IsZero() {
		now = clock.pausedAt
	}

	return now.Sub(clock.startedAt) - clock.pausedDuration
}

func (clock *GameClock) blinkOnAt(now time.Time) bool {
	return clock.elapsedAt(now)%goalBlinkInterval < goalBlinkOnDuration
}
